sentry-actix/src/lib.rs

```diff
@@ -191,21 +196,63 @@ where
 
     fn new_transform(&self, service: S) -> Self::Future {
         ok(SentryMiddleware {
-            service,
+            service: Rc::new(service),
             inner: self.clone(),
         })
     }
 }
 
 /// The middleware for individual services.
 pub struct SentryMiddleware<S> {
-    service: S,
+    service: Rc<S>,
     inner: Sentry,
 }
 
+async fn should_capture_request_body(headers: &HeaderMap) -> bool {
+    headers
+        .get("content-type")
+        .and_then(|h| h.to_str().ok())
+        .map(|s| s.to_lowercase())
+        .is_some_and(|ct| {
+            ct.contains("application/json") || ct.contains("application/x-www-form-urlencoded")
+        })
+}
+
+/// Extract a body from the HTTP request
+async fn body_from_http(req: &mut ServiceRequest) -> Result<BytesMut, PayloadError> {
+    let mut stream = req.take_payload();
+
+    let mut body = BytesMut::new();
+    while let Some(chunk) = futures_util::StreamExt::next(&mut stream).await {
+        let chunk = chunk?;
+        body.extend_from_slice(&chunk);
+    }
```

Comment:
I don’t think capturing the whole payload ahead of invoking the actual handler is not a good idea.

Although this is less likely for json / urlencoded payloads, the downstream middleware might expect to process the body in a streaming fashion. Plus, the body might be *huge*, and buffering it into memory might cause problems.
Again, this is highly unlikely to be a concern for json bodies though.

Reply:
that sounds reasonable